### PR TITLE
Work speed and work chance overrides

### PR DESCRIPTION
**Pull request by monochromatism**
## About The Pull Request

Adds two new procs (and associated variables) to abnormalities to allow them to alter work speed and success chance on the fly. Alters Schadenfreude to use the latter as a sort of demonstration.

## Why It's Good For The Game

More functionality enabling abnormalities to go beyond existing behavioral bounds, allowing both LC-accurate behavior for certain ones (including future ones) and more intricate behavior for those beyond the original LC.

## Changelog
:cl:
code: work speed and chance overrides
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/abnormality/he/schadenfreude.dm [hunk 1]
```diff
@@ -46,7 +47,7 @@
 	//Make sure there actually are two players on the Z level
 	var/living_players
 	for(var/mob/living/carbon/human/H in GLOB.player_list)
-		if(H.z == z && H.stat != DEAD)
```
[the remark this answers is not in the document]
Comment (monochromatism) [reply code/modules/mob/living/simple_animal/abnormality/he/schadenfreude.dm:49]: Kirie requested it be changed; she is the creator of the abnormality. This enables single-agent runs to not suffer the same consequences as true solo runs.

File: code/game/machinery/computer/abnormality_work.dm [hunk 2]
```diff
@@ -169,14 +171,18 @@
...
+		if(datum_reference.current.speed_override)
+			temp_work_speed = datum_reference.current.SpeedOverride(user, work_speed, temp_work_speed, work_type)
+		if(do_after(user, temp_work_speed, src, IGNORE_HELD_ITEM))
 			if(!CheckStatus(user))
 				break
 			user.remove_status_effect(/datum/status_effect/interventionshield) //removing status effect doesnt seem to effect all of parent. -IP
 			user.remove_status_effect(/datum/status_effect/interventionshield/white)
 			user.remove_status_effect(/datum/status_effect/interventionshield/black)
 			user.remove_status_effect(/datum/status_effect/interventionshield/pale)
-			if(do_work(work_chance))
```
[the remark this answers is not in the document]
Comment (monochromatism) [reply code/game/machinery/computer/abnormality_work.dm:179]: I _could_, but I want the behavior with override set FALSE to be _identical_ to the behavior before the introduction of overrides.

File: code/modules/mob/living/simple_animal/abnormality/he/schadenfreude.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (monochromatism) [reply code/modules/mob/living/simple_animal/abnormality/he/schadenfreude.dm:49]: Oh, alright then.

File: code/modules/mob/living/simple_animal/abnormality/_abnormality.dm [hunk 3]
```diff
@@ -53,6 +53,10 @@
 	var/work_damage_type = RED_DAMAGE
 	/// Maximum amount of PE someone can obtain per work procedure, if not null or 0.
 	var/max_boxes = null
+	// Work speed override. When true, SpeedOverride is called at the start of each work tick; when false, it's never called. Can be changed midwork.
+	var/speed_override = FALSE
+	// Work success chance override. When true, RateOverride is called at the start of each work tick; when false, it's never called. Can be changed midwork.
+	var/chance_override = FALSE
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/_abnormality.dm:59]: Obsolete. You don't need these.

File: code/modules/mob/living/simple_animal/abnormality/_abnormality.dm [hunk 4]
```diff
@@ -257,6 +261,17 @@
 /mob/living/simple_animal/hostile/abnormality/proc/AttemptWork(mob/living/carbon/human/user, work_type)
 	return TRUE
 
+// Overrides the normal work delay. Called in abnormality_work.dm each worktick while speed_override is TRUE.
+// work_speed is the vanilla value, temp_work_speed is the value as modified by any previous use of SpeedOverride.
+// If speed_override is TRUE and then set FALSE mid-work, the last value this returned will continue being used until the work ends. Useful for "flat" changes.
+// Remember, this is altering the FINAL value. Returning 0 makes an instant worktick, returning 20 makes a 2 second worktick.
+/mob/living/simple_animal/hostile/abnormality/proc/SpeedOverride(mob/living/carbon/human/user, work_speed, temp_work_speed, work_type)
```
Comment (EgorDinamit) [inline code/modules/mob/living/simple_animal/abnormality/_abnormality.dm:268]:
```suggestion
/mob/living/simple_animal/hostile/abnormality/proc/SpeedOverride(mob/living/carbon/human/user, work_speed, work_type)
```
We don't need new variable here or anywhere else.

**Review by EgorDinamit — CHANGES_REQUESTED** [review]
Get rid of "temporary" variables, they are not needed.

File: code/modules/mob/living/simple_animal/abnormality/_abnormality.dm [hunk 5]
```diff
@@ -265,6 +265,16 @@
 /mob/living/simple_animal/hostile/abnormality/proc/AttemptWork(mob/living/carbon/human/user, work_type)
 	return TRUE
 
+// Overrides the normal work delay. Called in abnormality_work.dm each worktick while speed_override is TRUE.
```
Comment (dsmith328) [inline code/modules/mob/living/simple_animal/abnormality/_abnormality.dm:268]:
```suggestion
// Overrides the normal work delay. Called in abnormality_work.dm each worktick.
```

**Review by dsmith328 — CHANGES_REQUESTED** [review]

**Review by dsmith328 — APPROVED** [review]

**Pull request merged**